FEATURE: Create upload_references table

nbianca opened the pull request:
This table holds associations between uploads and other models. This can be used to prevent removing uploads that are still in use.

<!-- NOTE: All pull requests should have tests (rspec in Ruby, qunit in JavaScript). If your code does not include test coverage, please include an explanation of why it was omitted. -->

tgxworld commented on app/models/upload_reference.rb:
```diff
@@ -0,0 +1,21 @@
+# Table name: upload_references
+#
+#  id          :bigint           not null, primary key
+#  upload_id   :bigint
+#  target_type :string
+#  target_id   :bigint
```
I think these columns should be `NOT NULL` too?

martin-brennan replied:
Yeah these definitely need to be NOT NULL, all of them

martin-brennan replied:
I think we should add `timestamps` to all tables we create too, it's a pain when they are not there when it comes to debugging issues.

martin-brennan commented on db/migrate/20220308210841_move_user_uploads_to_upload_references.rb:
```diff
@@ -0,0 +1,14 @@
+    SQL
+  end
+
+  def down
```
Should have `    raise ActiveRecord::IrreversibleMigration` here IMO

nbianca replied:
I do not think so, but I agree that we should have something here. If up copies data from a table to another one, down should copy that data back. In this case, that becomes difficult and unlikely to happen so that's why I left this blank.

martin-brennan replied:
If it's difficult and unlikely then the migration is irreversible and should be marked as such IMO, we follow this pattern all the time.

martin-brennan commented on db/migrate/20220309132719_move_post_uploads_to_upload_references.rb:
```diff
@@ -0,0 +1,14 @@
+    SQL
+  end
+
+  def down
```
Should have raise ActiveRecord::IrreversibleMigration here IMO

nbianca replied:
The same as above.

martin-brennan reviewed (commented):
Left a few comments, my main concern here is that we are dropping the `PostUpload` and `UserUpload` model and table in the same commit. We shouldn't do this until later on, there could be plugins relying on these, and also generally I don't think we should be deleting these old tables straight away until things have deployed and settled for a while and we're absolutely sure we don't need them.

I think we also may need `post_migrate` migrations to re-sync the PostUpload and UserUpload data to UploadReference, because records could be inserted into those tables during the deployment but after the original migration where we copy the data to UploadReference

ZogStriP commented on app/models/upload_reference.rb:
```diff
@@ -0,0 +1,61 @@
+  belongs_to :upload
+  belongs_to :target, polymorphic: true
+
+  def self.exist!(upload_ids: [], target: nil, target_type: nil, target_id: nil)
```
Not sure about that name 🤔. I find it a bit confusing because my brain always associates `exist` to a `test of presence`.

How about changing it to `ensure_exist!`? I find it better express the meaning on what this function does. What do you think?

nbianca replied:
That was the original name!

Then I considered that the exclamation mark at the end means "ensure" which meant that "ensure_exist!" meant "ensure exist ensure". I realized that "exist?" ends with a question mark anyway and the exclamation mark version would pair with it nice.

ZogStriP replied:
The `!` at the end of the method name doesn't mean "ensure". It means the method has side effects or is destructive.

nbianca replied:
In my mind, the destructive character is synonym to ensure. I renamed the method back.

ZogStriP commented on app/models/upload_reference.rb:
```diff
@@ -0,0 +1,61 @@
+    upload_ids = upload_ids.uniq.reject(&:blank?)
+    target_type = target_type.to_s
+
+    if upload_ids.empty?
```
I 👍 that you are avoiding to do any work when there are no references.

ZogStriP commented on app/models/site_setting.rb:
```diff
@@ -4,9 +4,22 @@ class SiteSetting < ActiveRecord::Base
+    if saved_change_to_value?
+      if self.data_type == SiteSettings::TypeSupervisor.types[:upload]
+        UploadReference.exist!(upload_ids: [self.value], target: self)
+      elsif self.data_type == SiteSettings::TypeSupervisor.types[:uploaded_image_list]
```
That case doesn't seem to be tested.

ZogStriP commented on app/models/user.rb:
```diff
@@ -148,6 +149,12 @@ class User < ActiveRecord::Base
 
+  after_save do
+    if saved_change_to_uploaded_avatar_id?
+      UploadReference.exist!(upload_ids: [self.uploaded_avatar_id], target: self)
```
Is this tested?

ZogStriP commented on db/migrate/20220330160747_copy_site_settings_uploads_to_upload_references.rb:
```diff
@@ -0,0 +1,16 @@
+      INSERT INTO upload_references(upload_id, target_type, target_id, created_at, updated_at)
+      SELECT value::integer, 'SiteSetting', id, created_at, updated_at
+      FROM site_settings
+      WHERE data_type = 18
```
How about the ones in `uploaded_image_list` ?

ZogStriP commented on app/models/upload.rb:
```diff
@@ -544,6 +509,25 @@ def self.migrate_to_new_scheme(limit: nil)
+      raw.scan(regexp).each do |match|
+        urls << match[0]
+        if match[1].present?
+          sha1s << match[1] if match[1] =~ /^[0-9a-fA-f]+$/
```
In [Ruby](https://ruby-doc.org/core-3.1.1/Regexp.html#class-Regexp-label-Character+Classes), you can shorten this regexp using the `\h` character class

```suggestion
          sha1s << match[1] if match[1] =~ /^\h+$/
```

ZogStriP commented on app/models/draft.rb:
```diff
@@ -82,7 +82,7 @@ def self.set(user, key, sequence, data, owner = nil, force_save: false)
       }
 
-      DB.exec(<<~SQL, opts)
+      draft_id, = DB.query_single(<<~SQL, opts)
```
Why the trailing comma?

nbianca replied:
DB.query_single returns an array with a single element. This line means that draft_id will store the first element.

ZogStriP replied:
Maybe do a first to indicate it's voluntary.

```suggestion
      draft_id = DB.query_single(<<~SQL, opts).first
```

ZogStriP commented on app/models/upload.rb:
```diff
@@ -544,6 +509,25 @@ def self.migrate_to_new_scheme(limit: nil)
     problems
   end
 
+  def self.extract_upload_ids(raw)
```
Doesn't look like this method is tested? I know it is indirectly because it's called in Draft, ReviewableQueuedPost and others, but having a test with a raw containing 1 URL matching each `InlineUploads.url_regexps` would ensure less regressions.

nbianca commented on db/post_migrate/20220309132720_copy_post_uploads_to_upload_references_for_sync.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class CopyPostUploadsToUploadReferencesForSync < ActiveRecord::Migration[6.1]
```
I added a comment.

ZogStriP replied:
Do we even need the first migration in the first place then? Do we need this for other tables as well?

ZogStriP commented on app/models/upload.rb:
```diff
@@ -546,6 +510,21 @@ def self.migrate_to_new_scheme(limit: nil)
+
+    sha1s = []
+
+    raw.scan(/upload:\/\/([a-zA-Z0-9]+)(?:\.[a-zA-Z0-9]+)?/).each do |match|
```
Why not use the same REGEXP as in https://github.com/discourse/discourse/blob/b96f90f9705a8f62222939fe3e613a215fdb2538/app/models/upload.rb#L337-L341 ?

nbianca replied:
There is a difference between the two regular expressions. The regular expression you propose is ideal when the whole string is a short URL. The regular expression I use is looking explicitly for `upload://`. I could probably make it shorter doing similar to that.

martin-brennan commented on app/models/post_upload.rb:
```diff
@@ -1,21 +0,0 @@
-# frozen_string_literal: true
```
Still don't think this model should be deleted in this PR, it is referred to in a few plugin places:

```
all-the-custom-plugins/plugins/discourse-warner-dc/db/fixtures/001_badges.rb
73:        INNER JOIN post_uploads pu

all-the-plugins/plugins/discourse-data-explorer/lib/queries.rb
351:    FROM post_uploads
352:    JOIN uploads ON uploads.id = post_uploads.upload_id
353:    JOIN posts ON posts.id = post_uploads.post_id

all-the-plugins/plugins/discourse-encrypt/spec/lib/email_sender_spec.rb
43:      reply.link_post_uploads
82:      reply.link_post_uploads

all-the-plugins/plugins/discourse-chat/spec/lib/chat_channel_archive_service_spec.rb
125:        expect(@channel_archive.destination_topic.posts.last.post_uploads.count).to eq(1)
```

We can just add a note that it's no longer used (pointing to `UploadReference`) and delete it when we drop the table later after plugins are changed.

martin-brennan commented on app/models/upload.rb:
```diff
@@ -91,43 +89,9 @@ def self.reset_in_use_callbacks
-    end
-
-    scope
+    self
```
Have we got specs that cover this change? It seems too simple compared to the original query but maybe it's just appearances and it's fine, just want to double check.

nbianca replied:
Yes, this part is very well covered. I did not add any new tests for this part because there is no change in functionality.

martin-brennan replied:
Okay cool just thought I would double check.

martin-brennan commented on app/models/upload_reference.rb:
```diff
@@ -0,0 +1,63 @@
+      target_id = target.id
+    end
+
+    upload_ids = upload_ids.uniq.reject(&:blank?)
```
I think `.compact` will work here rather than `.reject(&:blank?)` (unless maybe you are expecting these to sometimes be `""`?

nbianca replied:
I noticed that some invocations actually have IDs as strings and that is exactly why I made it this way. If I remember correctly, some tests fail with `.compact`.

martin-brennan replied:
No problem!

martin-brennan commented on app/models/upload_reference.rb:
```diff
@@ -0,0 +1,63 @@
+    target_type = target_type.to_s
+
+    if upload_ids.empty?
+      UploadReference
```
With this I feel like the `ensure_exists!` name is a little deceiving, because in this case it's deleting something. Maybe a `sync!` would be better? Or perhaps `upsert!`?

nbianca replied:
It used to be called just `exists!` and Regis suggested this name, which I like better. `upsert!` is also wrong because as you said, it is not just `update + insert`, but also `delete`. `sync!` sounds fine too.

martin-brennan commented on db/migrate/20220309132719_copy_post_uploads_to_upload_references.rb:
```diff
@@ -0,0 +1,16 @@
+      INSERT INTO upload_references(upload_id, target_type, target_id, created_at, updated_at)
+      SELECT upload_id, 'Post', post_id, NOW(), NOW()
+      FROM post_uploads
+      ON CONFLICT DO NOTHING
```
I am not sure if this will work, I am 95% sure that currently an upload can be linked to multiple posts, which has been an issue with secure media things. I don't think we can have this or the unique index on type/id/upload_id

martin-brennan commented on db/migrate/20220330160751_copy_badges_uploads_to_upload_references.rb:
```diff
@@ -0,0 +1,17 @@
+  def up
+    execute <<~SQL
+      INSERT INTO upload_references(upload_id, target_type, target_id, created_at, updated_at)
+      SELECT image_upload_id, 'Badge', id, created_at, updated_at
```
Yeah I generally tend to agree with @CvX here _except_ in the cases of link tables already existing, such as post_uploads. In that case we should use the created_at/updated_at from the link table.

martin-brennan commented on db/migrate/20220308201942_create_upload_references.rb:
```diff
@@ -0,0 +1,13 @@
+      t.timestamps
+    end
+
+    add_index :upload_references, [:upload_id, :target_type, :target_id], unique: true, name: 'index_upload_references_on_upload_and_target'
```
I am not sure if this will work, I am 95% sure that currently an upload can be linked to multiple posts, which has been an issue with secure media things. I am not sure if this is the case with other upload references, but yeah I don't think UNIQUE will work for this.

martin-brennan reviewed (changes requested):
A few questions and some concerns around the UNIQUE index, but thank you for addressing the points I made on my initial review. I think this is almost there, just a few things needed :+1:

martin-brennan commented:
Oh I just merged https://github.com/discourse/discourse/pull/16849 which has created conflicts for you here, I will fix them up :)

martin-brennan commented:
I originally made some comments about the UNIQUE index here then I realised 10 mins later they made no sense, so I now removed :sweat_smile:

nbianca commented:
There were many conflicts remaining so I fixed them and rebased the whole PR in the process to keep the history clean. I dropped your merge commit in the process. Hope you don't mind!

martin-brennan reviewed (approved):
I think this LGTM now, thanks for all your additional work after reviews!

davidtaylorhq commented on app/models/upload.rb:
```diff
@@ -542,6 +506,21 @@ def self.migrate_to_new_scheme(limit: nil)
     problems
   end
 
+  def self.extract_upload_ids(raw)
```
This method of extracting uploads from Drafts and Profiles is much more strict than the old system, and will not detect all upload URL types.

Maybe an easy fix is to make the new regex more general? Rather than searching for upload://, we could search for a / followed by 40 consecutive hex characters. Or a / followed by 27 Base62 characters. That would be more similar to the existing logic. Any false-positives would be handled by the Upload.where(sha: sha1s) lookup?

(see https://dev.discourse.org/t/29848/38)

nbianca replied:
I addressed this issue in the latest commit.
